File: services/panchangam/tithi_summary.go
package panchangam

import (
	"context"
	"fmt"
	"time"

	"github.com/naren-m/panchangam/astronomy"
	"github.com/naren-m/panchangam/astronomy/ephemeris"
	"github.com/naren-m/panchangam/observability"
	ppb "github.com/naren-m/panchangam/proto"
	"go.opentelemetry.io/otel/attribute"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

const (
	defaultTithiSummaryTimezone = "UTC"
	defaultTithiSummaryMethod   = "Drik"
	defaultTithiSummaryLocale   = "en"
	tithiBoundarySearchStep     = time.Hour
	tithiBoundarySearchWindow   = 72 * time.Hour
	tithiBoundaryPrecision      = time.Second
)

func (s *PanchangamServer) GetTithiSummary(ctx context.Context, req *ppb.GetTithiSummaryRequest) (*ppb.GetTithiSummaryResponse, error) {
	ctx, span := s.observer.CreateSpan(ctx, "PanchangamServer.GetTithiSummary")
	defer span.End()

	if req == nil {
		err := status.Error(codes.InvalidArgument, "request cannot be nil")
		span.RecordError(err)
		return nil, err
	}

	span.SetAttributes(
		attribute.String("request.at", req.At),
		attribute.Float64("request.latitude", req.Latitude),
		attribute.Float64("request.longitude", req.Longitude),
		attribute.String("request.timezone", req.Timezone),
		attribute.String("request.region", req.Region),
		attribute.String("request.method", req.CalculationMethod),
		attribute.String("request.locale", req.Locale),
		attribute.String("request.calendar_system", req.CalendarSystem),
	)

	input, err := validateTithiSummaryRequest(req)
	if err != nil {
		span.RecordError(err)
		return nil, err
	}

	span.SetAttributes(
		attribute.String("calculation.timezone", input.timezone),
		attribute.String("calculation.region", input.region),
		attribute.String("calculation.method", input.method),
		attribute.String("calculation.locale", input.locale),
		attribute.String("calculation.calendar_system", input.calendarSystem),
	)

	tithi, err := s.getTithiAt(ctx, input.at, input.calendarSystem)
	if err != nil {
		grpcErr := status.Error(codes.Internal, fmt.Sprintf("failed to calculate tithi: %v", err))
		observability.RecordError(ctx, grpcErr, observability.ErrorContext{
			Severity:  observability.SeverityHigh,
			Category:  observability.CategoryCalculation,
			Operation: "tithi_summary_current_tithi",
			Component: "panchangam_service",
		})
		span.RecordError(grpcErr)
		return nil, grpcErr
	}

	startTime, err := s.findTithiBoundary(ctx, input.at, tithi.Number, input.calendarSystem, -1)
	if err != nil {
		grpcErr := status.Error(codes.Internal, fmt.Sprintf("failed to calculate tithi start time: %v", err))
		span.RecordError(grpcErr)
		return nil, grpcErr
	}

	endTime, err := s.findTithiBoundary(ctx, input.at, tithi.Number, input.calendarSystem, 1)
	if err != nil {
		grpcErr := status.Error(codes.Internal, fmt.Sprintf("failed to calculate tithi end time: %v", err))
		span.RecordError(grpcErr)
		return nil, grpcErr
	}

	localAt := input.at.In(input.location)
	nakshatra, yoga, karana, vara, err := s.getPanchaAngaSummary(ctx, localAt, input.latitude, input.longitude)
	if err != nil {
		span.RecordError(err)
		return nil, err
	}

	day, err := getTithiSummaryDay(ctx, localAt, input.latitude, input.longitude, input.location)
	if err != nil {
		grpcErr := status.Error(codes.Internal, fmt.Sprintf("failed to calculate day details: %v", err))
		span.RecordError(grpcErr)
		return nil, grpcErr
	}

	resp := &ppb.GetTithiSummaryResponse{
		Date: localAt.Format("2006-01-02"),
		Tithi: &ppb.TithiSummary{
			Name:            tithi.Name,
			TraditionalName: tithi.TraditionalName,
			Number:          int32(tithi.Number),
			Paksha:          tithi.Paksha,
			PakshaDay:       int32(tithi.PakshaDay),
			Type:            string(tithi.Type),
			StartTime:       startTime.UTC().Format(time.RFC3339),
			EndTime:         endTime.UTC().Format(time.RFC3339),
		},
		PanchaAnga: &ppb.PanchaAngaSummary{
			Nakshatra: nakshatra.Name,
			Yoga:      yoga.Name,
			Karana:    karana.Name,
			Vara:      vara.GregorianDay,
		},
		Calculation: &ppb.TithiCalculationSummary{
			Timezone:       input.timezone,
			Region:         input.region,
			CalendarSystem: input.calendarSystem,
			Method:         input.method,
			Locale:         input.locale,
		},
		Day:           day,
		GeneratedAt:   input.at.UTC().Format(time.RFC3339),
		NextRefreshAt: endTime.UTC().Format(time.RFC3339),
	}

	span.AddEvent("Tithi summary response prepared", traceAttributes(
		"response.date", resp.Date,
		"response.tithi", resp.Tithi.Name,
		"response.sunrise", resp.Day.SunriseTime,
		"response.sunset", resp.Day.SunsetTime,
		"response.next_refresh_at", resp.NextRefreshAt,
	)...)

	return resp, nil
}

type tithiSummaryInput struct {
	at             time.Time
	latitude       float64
	longitude      float64
	timezone       string
	location       *time.Location
	region         string
	method         string
	locale         string
	calendarSystem string
}

func validateTithiSummaryRequest(req *ppb.GetTithiSummaryRequest) (*tithiSummaryInput, error) {
	if req.At == "" {
		return nil, status.Error(codes.InvalidArgument, "at parameter is required")
	}

	at, err := time.Parse(time.RFC3339, req.At)
	if err != nil {
		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid at timestamp: %v", err))
	}

	if req.Latitude < -90 || req.Latitude > 90 {
		return nil, status.Error(codes.InvalidArgument, "latitude must be between -90 and 90")
	}

	if req.Longitude < -180 || req.Longitude > 180 {
		return nil, status.Error(codes.InvalidArgument, "longitude must be between -180 and 180")
	}

	timezoneName := req.Timezone
	if timezoneName == "" {
		timezoneName = defaultTithiSummaryTimezone
	}

	location, err := NewTimezoneParser().ParseTimezone(timezoneName)
	if err != nil {
		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid timezone: %v", err))
	}

	method := req.CalculationMethod
	if method == "" {
		method = defaultTithiSummaryMethod
	}

	locale := req.Locale
	if locale == "" {
		locale = defaultTithiSummaryLocale
	}

	calendarSystem := req.CalendarSystem
	if calendarSystem == "" {
		calendarSystem = getCalendarSystemForRegion(req.Region)
	}
	if calendarSystem != "Purnimanta" && calendarSystem != "Amanta" {
		return nil, status.Error(codes.InvalidArgument, "calendar_system must be Purnimanta or Amanta")
	}

	return &tithiSummaryInput{
		at:             at,
		latitude:       req.Latitude,
		longitude:      req.Longitude,
		timezone:       timezoneName,
		location:       location,
		region:         req.Region,
		method:         method,
		locale:         locale,
		calendarSystem: calendarSystem,
	}, nil
}

func (s *PanchangamServer) getTithiAt(ctx context.Context, at time.Time, calendarSystem string) (*astronomy.TithiInfo, error) {
	jd := ephemeris.TimeToJulianDay(at)
	positions, err := s.ephemerisManager.GetPlanetaryPositions(ctx, jd)
	if err != nil {
		return nil, err
	}

	return s.tithiCalc.GetTithiFromLongitudesWithCalendarSystem(
		ctx,
		positions.Sun.Longitude,
		positions.Moon.Longitude,
		at,
		calendarSystem,
	)
}

func (s *PanchangamServer) findTithiBoundary(ctx context.Context, at time.Time, currentTithiNumber int, calendarSystem string, direction int) (time.Time, error) {
	if direction != -1 && direction != 1 {
		return time.Time{}, fmt.Errorf("direction must be -1 or 1")
	}

	currentSide := at.UTC()
	step := time.Duration(direction) * tithiBoundarySearchStep
	searchLimit := at.UTC().Add(time.Duration(direction) * tithiBoundarySearchWindow)

	for probe := currentSide.Add(step); withinBoundarySearch(probe, searchLimit, direction); probe = probe.Add(step) {
		probeTithi, err := s.getTithiAt(ctx, probe, calendarSystem)
		if err != nil {
			return time.Time{}, err
		}

		if probeTithi.Number != currentTithiNumber {
			if direction < 0 {
				return s.bisectTithiBoundary(ctx, probe, currentSide, currentTithiNumber, calendarSystem, false)
			}
			return s.bisectTithiBoundary(ctx, currentSide, probe, currentTithiNumber, calendarSystem, true)
		}

		currentSide = probe
	}

	return time.Time{}, fmt.Errorf("no tithi transition found within %s of %s", tithiBoundarySearchWindow, at.Format(time.RFC3339))
}

func withinBoundarySearch(probe, limit time.Time, direction int) bool {
	if direction < 0 {
		return !probe.Before(limit)
	}
	return !probe.After(limit)
}

func (s *PanchangamServer) bisectTithiBoundary(ctx context.Context, low, high time.Time, currentTithiNumber int, calendarSystem string, lowIsCurrent bool) (time.Time, error) {
	for high.Sub(low) > tithiBoundaryPrecision {
		mid := low.Add(high.Sub(low) / 2)
		midTithi, err := s.getTithiAt(ctx, mid, calendarSystem)
		if err != nil {
			return time.Time{}, err
		}

		if lowIsCurrent {
			if midTithi.Number == currentTithiNumber {
				low = mid
			} else {
				high = mid
			}
			continue
		}

		if midTithi.Number == currentTithiNumber {
			high = mid
		} else {
			low = mid
		}
	}

	return high.UTC().Truncate(time.Second), nil
}

func (s *PanchangamServer) getPanchaAngaSummary(ctx context.Context, at time.Time, latitude, longitude float64) (*astronomy.NakshatraInfo, *astronomy.YogaInfo, *astronomy.KaranaInfo, *astronomy.VaraInfo, error) {
	nakshatra, err := s.nakshatraCalc.GetNakshatraForDate(ctx, at)
	if err != nil {
		return nil, nil, nil, nil, status.Error(codes.Internal, fmt.Sprintf("failed to calculate nakshatra: %v", err))
	}

	yoga, err := s.yogaCalc.GetYogaForDate(ctx, at)
	if err != nil {
		return nil, nil, nil, nil, status.Error(codes.Internal, fmt.Sprintf("failed to calculate yoga: %v", err))
	}

	karana, err := s.karanaCalc.GetKaranaForDate(ctx, at)
	if err != nil {
		return nil, nil, nil, nil, status.Error(codes.Internal, fmt.Sprintf("failed to calculate karana: %v", err))
	}

	vara, err := s.varaCalc.GetVaraForDate(ctx, at, astronomy.Location{
		Latitude:  latitude,
		Longitude: longitude,
	})
	if err != nil {
		return nil, nil, nil, nil, status.Error(codes.Internal, fmt.Sprintf("failed to calculate vara: %v", err))
	}

	return nakshatra, yoga, karana, vara, nil
}

func getTithiSummaryDay(ctx context.Context, at time.Time, latitude, longitude float64, location *time.Location) (*ppb.DaySummary, error) {
	geo := astronomy.Location{
		Latitude:  latitude,
		Longitude: longitude,
	}

	sunTimes, err := astronomy.CalculateSunTimesWithContext(ctx, geo, at)
	if err != nil {
		return nil, err
	}

	traditionalPeriods, err := astronomy.CalculateTraditionalPeriodsWithContext(ctx, geo, at)
	if err != nil {
		return nil, err
	}
	if traditionalPeriods.AbhijitMuhurta == nil {
		return nil, fmt.Errorf("abhijit muhurta is unavailable")
	}

	return &ppb.DaySummary{
		SunriseTime: sunTimes.Sunrise.In(location).Format("15:04"),
		SunsetTime:  sunTimes.Sunset.In(location).Format("15:04"),
		AbhijitMuhurta: &ppb.TimeWindow{
			Name:       "Abhijit",
			StartTime:  traditionalPeriods.AbhijitMuhurta.Start.In(location).Format("15:04"),
			EndTime:    traditionalPeriods.AbhijitMuhurta.End.In(location).Format("15:04"),
			Auspicious: traditionalPeriods.AbhijitMuhurta.Auspicious,
		},
	}, nil
}

File: app/PanchangamShared/Sources/PanchangamShared/TithiSummary.swift
import Foundation

public enum TithiSummaryValidationError: Error, Equatable, LocalizedError, Sendable {
    case missingField(String)
    case invalidTithiNumber
    case invalidPakshaDay
    case invalidTithiWindow
    case invalidGeneratedWindow
    case invalidRefreshWindow
    case invalidTimezone

    public var errorDescription: String? {
        userMessage
    }

    public var userMessage: String {
        switch self {
        case .missingField(let field):
            return "\(field) is required"
        case .invalidTithiNumber:
            return "Tithi number must be between 1 and 30"
        case .invalidPakshaDay:
            return "Paksha day must be between 1 and 15"
        case .invalidTithiWindow:
            return "Tithi end time must be after start time"
        case .invalidGeneratedWindow:
            return "Generated time must be within the tithi start and end time"
        case .invalidRefreshWindow:
            return "Next refresh time must be after generated time and no later than tithi end time"
        case .invalidTimezone:
            return "Timezone is not recognized"
        }
    }
}

public struct TithiSummaryResponse: Codable, Equatable, Sendable {
    public var date: String
    public var tithi: TithiDetails
    public var panchaAnga: PanchaAngaSummary
    public var day: DaySummary
    public var calculation: TithiCalculationSummary
    public var generatedAt: Date
    public var nextRefreshAt: Date

    public init(
        date: String,
        tithi: TithiDetails,
        panchaAnga: PanchaAngaSummary,
        day: DaySummary,
        calculation: TithiCalculationSummary,
        generatedAt: Date,
        nextRefreshAt: Date
    ) {
        self.date = date
        self.tithi = tithi
        self.panchaAnga = panchaAnga
        self.day = day
        self.calculation = calculation
        self.generatedAt = generatedAt
        self.nextRefreshAt = nextRefreshAt
    }

    private enum CodingKeys: String, CodingKey {
        case date
        case tithi
        case panchaAnga = "pancha_anga"
        case day
        case calculation
        case generatedAt = "generated_at"
        case nextRefreshAt = "next_refresh_at"
    }

    public func validate() throws {
        try requireText(date, "Date")
        try tithi.validate()
        try panchaAnga.validate()
        try day.validate()
        try calculation.validate()

        guard generatedAt >= tithi.startTime, generatedAt < tithi.endTime else {
            throw TithiSummaryValidationError.invalidGeneratedWindow
        }

        guard nextRefreshAt > generatedAt, nextRefreshAt <= tithi.endTime else {
            throw TithiSummaryValidationError.invalidRefreshWindow
        }
    }

    private func requireText(_ value: String, _ field: String) throws {
        guard !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw TithiSummaryValidationError.missingField(field)
        }
    }
}

public struct TithiDetails: Codable, Equatable, Sendable {
    public var name: String
    public var traditionalName: String
    public var number: Int
    public var paksha: String
    public var pakshaDay: Int
    public var type: String
    public var startTime: Date
    public var endTime: Date

    public init(
        name: String,
        traditionalName: String,
        number: Int,
        paksha: String,
        pakshaDay: Int,
        type: String,
        startTime: Date,
        endTime: Date
    ) {
        self.name = name
        self.traditionalName = traditionalName
        self.number = number
        self.paksha = paksha
        self.pakshaDay = pakshaDay
        self.type = type
        self.startTime = startTime
        self.endTime = endTime
    }

    private enum CodingKeys: String, CodingKey {
        case name
        case traditionalName = "traditional_name"
        case number
        case paksha
        case pakshaDay = "paksha_day"
        case type
        case startTime = "start_time"
        case endTime = "end_time"
    }

    public func validate() throws {
        try requireText(name, "Tithi name")
        try requireText(traditionalName, "Traditional tithi name")
        try requireText(paksha, "Paksha")
        try requireText(type, "Tithi type")

        guard (1...30).contains(number) else {
            throw TithiSummaryValidationError.invalidTithiNumber
        }

        guard (1...15).contains(pakshaDay) else {
            throw TithiSummaryValidationError.invalidPakshaDay
        }

        guard endTime > startTime else {
            throw TithiSummaryValidationError.invalidTithiWindow
        }
    }

    private func requireText(_ value: String, _ field: String) throws {
        guard !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw TithiSummaryValidationError.missingField(field)
        }
    }
}

public struct PanchaAngaSummary: Codable, Equatable, Sendable {
    public var nakshatra: String
    public var yoga: String
    public var karana: String
    public var vara: String

    public init(nakshatra: String, yoga: String, karana: String, vara: String) {
        self.nakshatra = nakshatra
        self.yoga = yoga
        self.karana = karana
        self.vara = vara
    }

    public func validate() throws {
        try requireText(nakshatra, "Nakshatra")
        try requireText(yoga, "Yoga")
        try requireText(karana, "Karana")
        try requireText(vara, "Vara")
    }

    private func requireText(_ value: String, _ field: String) throws {
        guard !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw TithiSummaryValidationError.missingField(field)
        }
    }
}

public struct DaySummary: Codable, Equatable, Sendable {
    public var sunriseTime: String
    public var sunsetTime: String
    public var abhijitMuhurta: TimeWindow

    public init(sunriseTime: String, sunsetTime: String, abhijitMuhurta: TimeWindow) {
        self.sunriseTime = sunriseTime
        self.sunsetTime = sunsetTime
        self.abhijitMuhurta = abhijitMuhurta
    }

    private enum CodingKeys: String, CodingKey {
        case sunriseTime = "sunrise_time"
        case sunsetTime = "sunset_time"
        case abhijitMuhurta = "abhijit_muhurta"
    }

    public func validate() throws {
        try requireText(sunriseTime, "Sunrise")
        try requireText(sunsetTime, "Sunset")
        try abhijitMuhurta.validate()
    }

    private func requireText(_ value: String, _ field: String) throws {
        guard !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw TithiSummaryValidationError.missingField(field)
        }
    }
}

public struct TimeWindow: Codable, Equatable, Sendable {
    public var name: String
    public var startTime: String
    public var endTime: String
    public var auspicious: Bool

    public init(name: String, startTime: String, endTime: String, auspicious: Bool) {
        self.name = name
        self.startTime = startTime
        self.endTime = endTime
        self.auspicious = auspicious
    }

    private enum CodingKeys: String, CodingKey {
        case name
        case startTime = "start_time"
        case endTime = "end_time"
        case auspicious
    }

    public func validate() throws {
        try requireText(name, "Time window name")
        try requireText(startTime, "Time window start")
        try requireText(endTime, "Time window end")
    }

    private func requireText(_ value: String, _ field: String) throws {
        guard !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw TithiSummaryValidationError.missingField(field)
        }
    }
}

public struct TithiCalculationSummary: Codable, Equatable, Sendable {
    public var timezone: String
    public var region: String
    public var calendarSystem: String
    public var method: String
    public var locale: String

    public init(timezone: String, region: String, calendarSystem: String, method: String, locale: String) {
        self.timezone = timezone
        self.region = region
        self.calendarSystem = calendarSystem
        self.method = method
        self.locale = locale
    }

    private enum CodingKeys: String, CodingKey {
        case timezone
        case region
        case calendarSystem = "calendar_system"
        case method
        case locale
    }

    public func validate() throws {
        try requireText(timezone, "Timezone")
        try requireText(region, "Region")
        try requireText(calendarSystem, "Calendar system")
        try requireText(method, "Method")
        try requireText(locale, "Locale")

        guard TimeZone(identifier: timezone) != nil else {
            throw TithiSummaryValidationError.invalidTimezone
        }
    }

    private func requireText(_ value: String, _ field: String) throws {
        guard !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw TithiSummaryValidationError.missingField(field)
        }
    }
}

#if DEBUG
public extension TithiSummaryResponse {
    static var watchSimulatorPreview: TithiSummaryResponse {
        let generatedAt = Date()
        let startTime = generatedAt.addingTimeInterval(-6 * 60 * 60)
        let endTime = generatedAt.addingTimeInterval(6 * 60 * 60)

        return TithiSummaryResponse(
            date: "2026-06-03",
            tithi: TithiDetails(
                name: "Sashthi",
                traditionalName: "Sashthi",
                number: 6,
                paksha: "Shukla",
                pakshaDay: 6,
                type: "Nanda",
                startTime: startTime,
                endTime: endTime
            ),
            panchaAnga: PanchaAngaSummary(nakshatra: "Pushya", yoga: "Saubhagya", karana: "Taitila", vara: "Wednesday"),
            day: DaySummary(
                sunriseTime: "05:42",
                sunsetTime: "19:01",
                abhijitMuhurta: TimeWindow(name: "Abhijit", startTime: "11:54", endTime: "12:48", auspicious: true)
            ),
            calculation: TithiCalculationSummary(
                timezone: "America/Los_Angeles",
                region: "California",
                calendarSystem: "Purnimanta",
                method: "Drik",
                locale: "en"
            ),
            generatedAt: generatedAt,
            nextRefreshAt: generatedAt.addingTimeInterval(60 * 60)
        )
    }
}
#endif
